### PR TITLE
feat(jwt) add support to a custom header name

### PR DESCRIPTION
**Pull request by raelmax**
### Summary

It allow the jwt plugin to work together with another authentication
method that uses the 'Authorization' header

Reference: https://discuss.konghq.com/t/custom-header-on-jwt-plugin/1524

### Full changelog

* Add a `header_names` field to jwt plugin schema;
* Change the `retrieve_token` method to  use the `header_names` field;
* Add a test to check if the custom `header_names` is used;
* Bump jwt plugin version from `0.2.0` to `0.2.1`.

### CODE REVIEW

File: kong/plugins/jwt/schema.lua [hunk 1]
```diff
@@ -28,6 +28,7 @@ return {
     anonymous = {type = "string", default = "", func = check_user},
     run_on_preflight = {type = "boolean", default = true},
     maximum_expiration = {type = "number", default = 0, func = check_positive},
+    header_name = {type = "string", default = "authorization"},
```
Comment (Tieske) [inline kong/plugins/jwt/schema.lua]: both the `uri_param_names` and the `cookie_names` accept an array. So it makes sense to also add `header_names` as an array

File: kong/plugins/jwt/handler.lua [hunk 2]
```diff
@@ -13,11 +13,11 @@ local get_method     = ngx.req.get_method
 local JwtHandler = BasePlugin:extend()
 
 JwtHandler.PRIORITY = 1005
-JwtHandler.VERSION = "0.1.0"
+JwtHandler.VERSION = "0.1.1"
 
 --- Retrieve a JWT in a request.
 -- Checks for the JWT in URI parameters, then in cookies, and finally
--- in the `Authorization` header.
+-- in the configured header_name(defaults to `Authorization`).
```
Comment (Tieske) [inline kong/plugins/jwt/handler.lua]: small typo, space before `(`

File: kong/plugins/jwt/handler.lua [hunk 3]
```diff
@@ -39,7 +39,7 @@ local function retrieve_token(request, conf)
     end
   end
 
-  local authorization_header = request.get_headers()["authorization"]
+  local authorization_header = request.get_headers()[conf.header_name or "authorization"]
```
Comment (Tieske) [inline kong/plugins/jwt/handler.lua]: please rebase on `next` instead of `master`, and then skip setting the default here, since those will be automatically set [since this pr](https://github.com/Kong/kong/pull/3756).

**Review by Tieske — CHANGES_REQUESTED** [review]
Thx for your contribution @raelmax. Comments are some minor things, and a change to an array.

**Comment by raelmax** [comment]
@Tieske I've fixed the minor changes and change the field to an array. Since I've rebased this branch on `next`, should I reopen this pull request to the `next` branch?

**Comment by Tieske** [comment]
@raelmax at the top, right of the PR title is an `edit` button. It allows you to change the target branch. I already changed it to `next`.

File: kong/plugins/jwt/handler.lua [hunk 4]
```diff
@@ -38,20 +38,25 @@ local function retrieve_token(request, conf)
...
-  local authorization_header = request.get_headers()["authorization"]
-  if authorization_header then
-    local iterator, iter_err = ngx_re_gmatch(authorization_header, "\\s*[Bb]earer\\s+(.+)")
-    if not iterator then
-      return nil, iter_err
-    end
+  local request_headers = request.get_headers()
+  for _, v in ipairs(conf.header_names) do
+    local token_header = request_headers[v]
+    if token_header then
```
Comment (Tieske) [inline kong/plugins/jwt/handler.lua]: this value `token_header` can potentially not be a string, but a table. This happens when the same header occurs multiple times.

Add something like:
```lua
if type(token_header) == "table" then
  token_header = token_header[1]
end
```

would be nice to add a regression test for this as well (not failing on duplicate headers)

**Comment by Tieske** [comment]
Hmmm just noticed CI seems to be gone from this PR... Maybe better to indeed close this and open a new PR. (not sure that was the cause, but can't think of anything else)

File: kong/plugins/jwt/handler.lua (hunk 4 above)
Comment (raelmax) [reply kong/plugins/jwt/handler.lua]: @Tieske, sorry for the delay. I add the test and fix the case with a table instead a string. Should I reopen the merge request and close this one?

**Comment by raelmax** [comment]
Should I close this PR and open another one?

**Comment by thibaultcha** [comment]
@raelmax Please keep the same PR.m, but rebase your patch on top of the next branch and force push your changes.

**Comment by raelmax** [comment]
Hey guys, any update about the merge of this pull request? =)

**Comment by thibaultcha** [comment]
@raelmax Hi,

At the moment we are busy with the 1.0 release, and we are delaying all unrelated work (including feature PRs). Sorry about the delay, once 1.0 out, we'll get back to this PR. Thanks.

**Comment by oovolkov** [comment]
Hi guys.
Are there any plans to move forward with this PR.
This feature is highly required.
Thank you in advance.

File: kong/plugins/jwt/handler.lua [hunk 5]
```diff
@@ -41,20 +41,28 @@ local function retrieve_token(conf)
...
+  local request_headers = kong.request.get_headers()
+  for _, v in ipairs(conf.header_names) do
+    local token_header = request_headers[v]
+    if token_header then
+      if type(token_header) == "table" then
+        token_header = token_header[1]
+      end
+      local iterator, iter_err = re_gmatch(token_header, "\\s*[Bb]earer\\s+(.+)")
+      if not iterator then
+        ngx.log(ngx.ERR, "[jwt]: ", iter_err)
```
Comment (hishamhm) [inline kong/plugins/jwt/handler.lua]: needs to be replaced with `kong.log.err`

File: kong/plugins/jwt/handler.lua [hunk 6]
```diff
@@ -41,20 +41,28 @@ local function retrieve_token(conf)
...
+        break
+      end
 
-    local m, err = iterator()
-    if err then
-      return nil, err
-    end
+      local m, err = iterator()
+      if err then
+        ngx.log(ngx.ERR, "[jwt]: ", err)
```
Comment (hishamhm) [inline kong/plugins/jwt/handler.lua]: needs to be replaced with `kong.log.err`

**Review by hishamhm — COMMENTED** [review]
Hi @raelmax, sorry for not getting back to this! The PR currently has conflicts, but I think a simple rebase should fix it and get it in a approvable/mergeable state!

File: kong/plugins/jwt/handler.lua (hunk 6 above)
Comment (oovolkov) [reply kong/plugins/jwt/handler.lua]: Hi guys
Who can fix this PR?
Thank you in advance

**Comment by Tieske** [comment]
replaced by #4757

**Comment by gszr** [comment]
@raelmax, Thank you for your contribution and apologies for the delay on our end. It has been merged (thanks @Tieske) and will be in Kong 1.3.
